src/clusterfuzz/_internal/bot/fuzzers/centipede/engine.py

```diff
@@ -0,0 +1,219 @@
+  if not centipede_path.exists():
+    raise CentipedeError('Centipede not found in build')
+
+  centipede_path = str(centipede_path)
+  if environment.get_value('USE_UNSHARE'):
+    return new_process.UnicodeModifierRunner(centipede_path)
+  return new_process.UnicodeProcessRunner(centipede_path)
+
+
+def _get_reproducer_path(log):
+  """Gets the reproducer path, if any."""
+  crash_match = _CRASH_REGEX.search(log)
+  if not crash_match:
+    return None
+
+  return crash_match.group(1)
+
+
+class Engine(engine.Engine):
+  """Centipede engine implementation."""
+
+  @property
+  def name(self):
+    return 'centipede'
+
+  def prepare(self, corpus_dir, target_path, build_dir):  # pylint: disable=unused-argument
+    """Prepares for a fuzzing session, by generating options.
+
+    Args:
+      corpus_dir: The main corpus directory.
+      target_path: Path to the target.
+      build_dir: Path to the build directory.
+
+    Returns:
+      A FuzzOptions object.
+    """
+    arguments = []
+    dict_path = Path(
+        dictionary_manager.get_default_dictionary_path(target_path))
+    if dict_path.exists():
+      arguments.append(f'--dictionary={dict_path}')
+
+    # Directory workdir saves:
+    # 1. Centipede-readable corpus file;
+    # 2. Centipede-readable feature file;
+    # 3. Crash reproducing inputs.
+    workdir = Path(corpus_dir).parent / 'workdir'
+    self._create_temp_dir(workdir)
+    arguments.append(f'--workdir={workdir}')
+
+    # Directory corpus_dir saves the corpus files required by ClusterFuzz.
+    self._create_temp_dir(corpus_dir)
+    arguments.append(f'--corpus_dir={corpus_dir}')
+
+    # The unsanitized binary, Centipede requires it to be the main fuzz target.
+    arguments.append(f'--binary={target_path}')
+
+    # Extra sanitized binaries, Centipede requires to build them separately.
+    # Assuming they will be in child dirs named by fuzzer_utils.EXTRA_BUILD_DIR.
+    sanitized_target_name = Path(target_path).name
+    sanitized_target_path = Path(build_dir, fuzzer_utils.EXTRA_BUILD_DIR,
+                                 sanitized_target_name)
+    if sanitized_target_path.exists():
+      arguments.append(f'--extra_binaries={sanitized_target_path}')
+
+    return engine.FuzzOptions(corpus_dir, arguments, {})
+
+  def fuzz(self, target_path, options, reproducers_dir, max_time):  # pylint: disable=unused-argument
+    """Runs a fuzz session.
+
+    Args:
+      target_path: Path to the target.
+      options: The FuzzOptions object returned by prepare().
+      reproducers_dir: The directory to put reproducers in when crashes
+          are found.
+      max_time: Maximum allowed time for the fuzzing to run.
+
+   Returns:
+      A FuzzResult object.
+    """
+    runner = _get_runner()
+    arguments = _DEFAULT_ARGUMENTS.copy()
+    arguments.extend(options.arguments)
+
+    timeout = max_time + _CLEAN_EXIT_SECS
+    fuzz_result = runner.run_and_wait(
+        additional_args=arguments, timeout=timeout)
```

Comment:
We should be propagating the timeout (without _CLEAN_EXIT_SECS) to the centipede commandline to let it exit on its own.

Reply:
Sorry, I am not sure if I understand this correctly:
Would you prefer to pass `max_time` as a command line params (e.g., `centipede --timeout=max_time <centipede args>`)?
`Centipede` does not have a parameter to limit the overall fuzzing time, though. The only relevant flag [`timeout`](https://github.com/google/centipede/blob/43a5a71bbfbb83f5c3f6df7e5c0ede57d238d3d5/environment.cc#L95) is about the execution time of one input.
Or would you prefer to use the [`timeout`](https://man7.org/linux/man-pages/man1/timeout.1.html) command (e.g., `timeout $max_time centipede <centipede args>`)?

Reply:
OK, I forgot that Centipede doesn't have this flag. Let's skip this for now, and see if it causes any issues.